src/metabase/lib/query.cljc

```diff
@@ -433,3 +434,73 @@
   (-> a-query
       (dissoc a-query :lib/metadata)
       lib.cache/discard-query-cache))
+
+(defn- stage-seq* [dataset-query]
+  (cond
+    (vector? dataset-query)
+    (case (first dataset-query)
+      :metric
+      [{:source-card (get dataset-query 2)}]
+
+      (mapcat stage-seq* dataset-query))
+
+    (map? dataset-query)
+    (concat (:stages dataset-query) (mapcat stage-seq* (vals dataset-query)))
+
+    :else
+    []))
+
+(defn- stage-seq [card-id dataset-query]
+  (map #(assoc % ::from-card card-id) (stage-seq* dataset-query)))
+
+(defn- expand-stage [metadata-provider stage]
+  (let [card-id (:source-card stage)
+        card (when card-id (lib.metadata/card metadata-provider card-id))
+        expanded-query (some->> card
+                                :dataset-query
+                                (query metadata-provider))]
```

Comment:
This still needs normalization, like in `dataset-query->query`. The legacy MBQL might be really outdated and not understood by our logic. If we don't have a helper function in the lib that will go from `:metadata/card` to a properly formed pMBQL query, we 100% should.

(Precisely because it's tricky and easy to get wrong! Another example wrinkle: on the FE we can assume it's already been normalized by the BE before being sent over.)

This can be tackled in a separate PR, though - just add the explicit normalize call here for now.

Reply:
Oof, that created a cyclic dependency. 

`metabase.models.card.metadata` depends on `metabase.lib.core`.

Any ideas how to resolve it?

Reply:
https://github.com/metabase/metabase/blob/35980-redux-circular-dependencies/src/metabase/models/card/metadata.clj#L72